File: app/bloom_care/lib/app-backend-code/services/flower_mapper.dart
import '../models/daily_summary.dart';
import '../models/flower_parameters.dart';

class FlowerMapper {
  const FlowerMapper();

  FlowerParameters map(
    DailySummary summary, {
    double growthHeight = 0,
  }) {
    final hueShift = _normalize(
      summary.averageSkinTemp,
      inputMin: 36.3,
      inputMax: 36.8,
      outputMin: -20,
      outputMax: 20,
    );

    final saturation = _normalize(
      summary.averageSpo2,
      inputMin: 90,
      inputMax: 100,
      outputMin: 0.55,
      outputMax: 1,
    );

    // In this prototype, activity means movement during sleep.
    // More movement therefore produces a less open flower.
    final petalOpenness =
        (1 - (summary.averageActivity / 100).clamp(0, 1)).toDouble();

    return FlowerParameters(
      hueShiftDegrees: hueShift,
      saturation: saturation,
      petalOpenness: petalOpenness,
      pulseBpm: summary.averageHeartbeat,
      growthHeight: growthHeight.clamp(0, 1).toDouble(),
    );
  }

  double _normalize(
    double value, {
    required double inputMin,
    required double inputMax,
    required double outputMin,
    required double outputMax,
  }) {
    final ratio =
        ((value - inputMin) / (inputMax - inputMin)).clamp(0, 1).toDouble();
    return outputMin + ratio * (outputMax - outputMin);
  }
}

File: web-prototype/src/sensor/flowerMapper.js
// Mirrors app/bloom_care/lib/app-backend-code/services/flower_mapper.dart
// Same normalization ranges, so the numbers here should match what the Flutter
// backend produces for the same sensor inputs once it's wired to real hardware.

function normalize(value, inputMin, inputMax, outputMin, outputMax) {
  const ratio = Math.min(1, Math.max(0, (value - inputMin) / (inputMax - inputMin)));
  return outputMin + ratio * (outputMax - outputMin);
}

export function mapToFlowerParameters(summary, growthHeight = 0) {
  const hueShiftDegrees = normalize(summary.averageSkinTemp, 36.3, 36.8, -20, 20);
  const saturation = normalize(summary.averageSpo2, 90, 100, 0.55, 1);

  // Same as the Dart version's comment: activity here means movement, so more
  // movement produces a less "open" flower.
  const petalOpenness = 1 - Math.min(1, Math.max(0, summary.averageActivity / 100));

  return {
    hueShiftDegrees,
    saturation,
    petalOpenness,
    pulseBpm: summary.averageHeartbeat,
    growthHeight: Math.min(1, Math.max(0, growthHeight)),
  };
}
